mlir/lib/Conversion/TosaToTensor/TosaToTensor.cpp

```diff
@@ -127,81 +127,74 @@ static bool createReassociationMapsForCollapse(
 }
```

Comment:
@ljfitz, please check the the changes I had to do here after your review to fix test failures.
Essentially, ReshapeConverterCollapseExpand was creating a sequence of foldable reshapes and hoped that other pattern in here (ReshapeConverterCollapse  & ReshapeConverterExpand) would catch them.
Now we see them folded before the patterns apply.
I solved this by merging ReshapeConverterCollapse and ReshapeConverterExpand into ReshapeConverterCollapseExpand, so there is only a single pattern that does the job.

Reply:
My head hurts a bit trying to work out if ReshapeConverterCollapse & ReshapeConverterExpand had value independent of ReshapeConverterCollapseExpand but I suspect you are right that they were only there to support the latter. The changes look good.